companion/src/mainwindow.cpp

```diff
@@ -1336,16 +1352,36 @@ void MainWindow::writeFlash(QString fileToFlash)
     g.backupOnFlash(backup);
     if (backup) {
       QString backupFile = generateProcessUniqueTempFileName("backup.bin");
-      bool backupEnable=g.enableBackup();
-      QString backupPath=g.backupDir();
-      if (!backupPath.isEmpty() && !IS_TARANIS(GetEepromInterface()->getBoard())) {
-        if (!QDir(backupPath).exists()) {
-          if (backupEnable) {
-            QMessageBox::warning(this, tr("Backup is impossible"), tr("The backup dir set in preferences does not exist"));
+      QString backupPath;
+      bool backupEnable = g.profile[g.id()].penableBackup();
+      if (backupEnable) {
+        backupPath=g.profile[g.id()].pBackupDir();
+        if (!backupPath.isEmpty()) {
+          if (!QDir(backupPath).exists()) {
+            backupEnable = false;
+          }
+        } 
+        else {
+          backupEnable = false;
+        }        
+      }
+      if (!backupEnable) {
+        backupEnable=g.enableBackup();
+        backupPath = g.backupDir();
+        if (!backupPath.isEmpty()) {
+          if (!QDir(backupPath).exists()) {
+            backupEnable = false;
           }
-          backupEnable=false;
+        }
+        else {
+          backupEnable = false;
         }
       }
+      if ((g.enableBackup() || g.profile[g.id()].penableBackup()) && !(backupEnable) && !IS_TARANIS(GetEepromInterface()->getBoard())) {
+        if (!QDir(backupPath).exists()) {
+          QMessageBox::warning(this, tr("Backup is impossible"), tr("The backup dir set in preferences does not exist"));
+        }      
+      } 
       else {
```

Comment:
Can we put this section and one in MainWindow::writeBackup() in a new function? They look very similar.